File: 4. Median of Two Sorted Arrays/median_of_two_sorted_arrays.js
function findMedianSortedArrays(nums1, nums2) {
    if (nums1.length > nums2.length) {
        [nums1, nums2] = [nums2, nums1];
    }

    const m = nums1.length;
    const n = nums2.length;
    const total = m + n;
    const half = Math.floor(total / 2);

    let left = 0;
    let right = m;

    while (true) {
        let i = Math.floor((left + right) / 2);
        let j = half - i;

        const left1 = i > 0 ? nums1[i - 1] : -Infinity;
        const right1 = i < m ? nums1[i] : Infinity;
        const left2 = j > 0 ? nums2[j - 1] : -Infinity;
        const right2 = j < n ? nums2[j] : Infinity;

        if (left1 <= right2 && left2 <= right1) {
            if (total % 2 === 0) {
                return (
                    (Math.max(left1, left2) + Math.min(right1, right2)) / 2
                );
            } else {
                return Math.min(right1, right2);
            }
        } else if (left1 > right2) {
            right = i - 1;
        } else {
            left = i + 1;
        }
    }
}

function main() {
    console.log(findMedianSortedArrays([1, 3], [2])); // 2.0
    console.log(findMedianSortedArrays([1, 2], [3, 4])); // 2.5
    console.log(findMedianSortedArrays([0, 0], [0, 0])); // 0
}

main();

File: 4. Median of Two Sorted Arrays/median_of_two_sorted_arrays.py
def findMedianSortedArrays(nums1, nums2):
    if len(nums1) > len(nums2):
        nums1, nums2 = nums2, nums1

    m, n = len(nums1), len(nums2)
    total = m + n
    half = total // 2

    left, right = 0, m

    while True:
        i = (left + right) // 2
        j = half - i

        left_nums1  = nums1[i - 1] if i > 0 else float("-inf")
        right_nums1 = nums1[i]     if i < m else float("inf")
        left_nums2  = nums2[j - 1] if j > 0 else float("-inf")
        right_nums2 = nums2[j]     if j < n else float("inf")

        if left_nums1 <= right_nums2 and left_nums2 <= right_nums1:
            if total % 2 == 0:
                return (max(left_nums1, left_nums2) + min(right_nums1, right_nums2)) / 2
            else:
                return min(right_nums1, right_nums2)
        elif left_nums1 > right_nums2:
            right = i - 1
        else:
            left = i + 1

if __name__ == "__main__":
    nums1 = [1, 3]
    nums2 = [2]
    print(findMedianSortedArrays(nums1, nums2))  # 2.0

    nums1 = [1, 2]
    nums2 = [3, 4]
    print(findMedianSortedArrays(nums1, nums2))  # 2.5

    nums1 = [0, 0]
    nums2 = [0, 0]
    print(findMedianSortedArrays(nums1, nums2))  # 0
